Mainnet/STOContract/STOContract/STOContract.cs

```diff
@@ -0,0 +1,1361 @@
+    {
+        get => this.PersistentState.GetUInt64(nameof(EndBlock));
+        private set => this.PersistentState.SetUInt64(nameof(EndBlock), value);
+    }
+
+    public Address TokenAddress
+    {
+        get => this.PersistentState.GetAddress(nameof(TokenAddress));
+        private set => this.PersistentState.SetAddress(nameof(TokenAddress), value);
+    }
+
+    public Address KYCAddress
+    {
+        get => this.PersistentState.GetAddress(nameof(KYCAddress));
+        private set => this.PersistentState.SetAddress(nameof(KYCAddress), value);
+    }
+
+    public Address MapperAddress
+    {
+        get => this.PersistentState.GetAddress(nameof(MapperAddress));
+        private set => this.PersistentState.SetAddress(nameof(MapperAddress), value);
+    }
+
+    public ulong TokenBalance
+    {
+        get => this.PersistentState.GetUInt64(nameof(TokenBalance));
+        private set => this.PersistentState.SetUInt64(nameof(TokenBalance), value);
+    }
+
+    public bool IsNonFungibleToken
+    {
+        get => this.PersistentState.GetBool(nameof(IsNonFungibleToken));
+        private set => this.PersistentState.SetBool(nameof(IsNonFungibleToken), value);
+    }
+
+    public bool SaleOpen => EndBlock >= this.Block.Number && TokenBalance > 0;
+
+    public Address Owner
+    {
+        get => this.PersistentState.GetAddress(nameof(Owner));
+        private set => this.PersistentState.SetAddress(nameof(Owner), value);
+    }
+
+    public SalePeriod[] SalePeriods
+    {
+        get => PersistentState.GetArray<SalePeriod>(nameof(SalePeriods));
+        private set => PersistentState.SetArray(nameof(SalePeriods), value);
+    }
+
+    public STOContract(ISmartContractState smartContractState,
+                       Address owner,
+                       uint tokenType,
+                       ulong totalSupply,
+                       string name,
+                       string symbol,
+                       Address kycAddress,
+                       Address mapperAddress,
+                       byte[] salePeriods) : base(smartContractState)
+    {
+        Assert(tokenType < 3, $"The {nameof(tokenType)} parameter can be between 0 and 2");
+
+        Assert(PersistentState.IsContract(kycAddress), "The kycAdress is not a contract adress");
+
+        var periods = Serializer.ToArray<SalePeriodInput>(salePeriods);
+
+        ValidatePeriods(periods);
+        var tokenTypeEnum = (TokenType)tokenType;
+        var result = CreateTokenContract(tokenTypeEnum, totalSupply, name, symbol);
+
+        Assert(result.Success, "Creating token contract failed.");
+
+        Log(new STOSetupLog { TokenAddress = result.NewContractAddress });
+
+        KYCAddress = kycAddress;
+        MapperAddress = mapperAddress;
+        TokenAddress = result.NewContractAddress;
+        IsNonFungibleToken = tokenTypeEnum == TokenType.NonFungibleToken;
+        TokenBalance = IsNonFungibleToken ? ulong.MaxValue : totalSupply;
+        Owner = owner;
```

Comment:
KycAddress has been validated to be a contract, but we might want to check that `owner` and `mapperAddress` are not `Address.Zero`.

Reply:
Contract call will be failed in the case of address parameter. It can be zero address on solidity if parameter is not provided. 

I follow StandardToken example and addresses are not checked for default values.

Reply:
So you're saying that sending an Address.Zero address will fail the transaction anyway?